src/batMap.ml

```diff
@@ -495,7 +577,7 @@ module Concrete = struct
   let filterv f t cmp =
     foldi (fun k a acc -> if f a then add k a cmp acc else acc) t empty
   let filter f t cmp =
-    foldi (fun k a acc -> if f k a then add k a cmp acc else acc) t empty
+    foldi (fun k a acc -> if f k a then acc else remove k cmp acc) t t
```

Comment:
Why did you change the implementation here? If I understand correctly, the previous implementation is better when few keys remain in the result, and the new implementation is better when most keys remain in the result.

Remarks:
- it is odd that `filterv` and `filter` would use opposite algorithms
- we could implement a direct version using `bal` that should be fast in all cases (but then we have already spent a lot of time on this PR so maybe it's okay to keep less-good code if it was already there).

Reply:
Because if the callback `f` returns true for all bindings then the given map should be returned unmodified.
That behaviour cannot be achieved by starting from `empty` and adding the missing elements...

I can make the implementation of filterv and filter consistent though.

Reply:
I may rewrite `filter_map` to use ordered join operations instead of this unpleasant `fold`, and define `filter` and `filterv` on top of it. This can wait until the PR is merged.